### PR TITLE
VMware: vmware_vm_inventory performance improvement

### LINKED ISSUE #56786: VMware: vmware_vm_inventory Won't Complete With Large Inventory
##### SUMMARY

We have a fairly large vSphere 6.5 environment with about 6600 VMs. When kicking off an ansible-inventory command with vmware_vm_inventory, I can't get it to complete, have waited more than 20 minutes on some tests. I can't tell if the issue is due to the number of VMs, or a different issue. Have also tried `with_tags: no`
##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

vmware_vm_inventory
##### ANSIBLE VERSION

```paste below
ansible 2.8.0
  config file = /home/useracct/.ansible.cfg
  configured module search path = [u'/home/useracct/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/useracct/.local/lib/python2.7/site-packages/ansible
  executable location = /home/useracct/.local/bin/ansible
  python version = 2.7.5 (default, Apr  9 2019, 14:30:50) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]
```

##### CONFIGURATION

```paste below
DEFAULT_HASH_BEHAVIOUR(/home/useracct/.ansible.cfg) = merge
DEFAULT_LOAD_CALLBACK_PLUGINS(/home/useracct/.ansible.cfg) = True
DEFAULT_ROLES_PATH(/home/useracct/.ansible.cfg) = [u'/home/useracct/src/ansible-ops/.imported_roles', u'/home/useracct/roles']
DEFAULT_TIMEOUT(/home/useracct/.ansible.cfg) = 30
DEFAULT_VAULT_PASSWORD_FILE(env: ANSIBLE_VAULT_PASSWORD_FILE) = /home/useracct/.ansible/vaultpa
HOST_KEY_CHECKING(/home/useracct/.ansible.cfg) = False
INVENTORY_ENABLED(/home/useracct/.ansible.cfg) = [u'host_list', u'script', u'auto', u'yaml', u'ini', u'toml', u'vmware_vm_inventory']
```

##### OS / ENVIRONMENT

Target vsphere is 6.5

##### STEPS TO REPRODUCE

```yaml
plugin: vmware_vm_inventory
strict: False
hostname: <host>
username: <username>
password: <password>
validate_certs: False
with_tags: True
```

##### EXPECTED RESULTS

Expected to have a return in 5-10 seconds.

##### ACTUAL RESULTS

Let this one run 18 minutes before stopping.

```paste below
[useracct@host ansible-inv]$ time ansible-inventory -vvvv --list -i vmware.yml
ansible-inventory 2.8.0
  config file = /home/useracct/.ansible.cfg
  configured module search path = [u'/home/useracct/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/useracct/.local/lib/python2.7/site-packages/ansible
  executable location = /home/useracct/.local/bin/ansible-inventory
  python version = 2.7.5 (default, Apr  9 2019, 14:30:50) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]
Using /home/useracct/.ansible.cfg as config file
setting up inventory plugins
host_list declined parsing /home/useracct/ansible-inv/vmware.yml as it did not pass it's verify_file() method
script declined parsing /home/useracct/ansible-inv/vmware.yml as it did not pass it's verify_file() method
Not replacing invalid character(s) "set(['-'])" in group name (Non-managed)
[DEPRECATION WARNING]: The TRANSFORM_INVALID_GROUP_CHARS settings is set to allow bad characters in group names by default, this will change, but still be user configurable on deprecation. This feature will be removed in version 2.10. Deprecation warnings
can be disabled by setting deprecation_warnings=False in ansible.cfg.
 [WARNING]: Invalid characters were found in group names but not replaced, use -vvvv to see details

Not replacing invalid character(s) "set([' '])" in group name (Build Year 2018)
Not replacing invalid character(s) "set(['.'])" in group name (com.vmware.vr.HasVrDisks)

^C [ERROR]: User interrupted execution


real	18m3.745s
user	10m6.323s
sys	0m7.720s
```

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY

Fixes: #56786

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE 
- Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/plugins/inventory/vmware_vm_inventory.py

### CODE REVIEW

**Comment by Akasurde** [comment]
@gparent Could you please try this PR and let me know? Thanks in advance.

For me - 

Before change
```
# time ansible-inventory -i vmware_vm_inventory.yml --list
...
2.29s user 0.21s system 1% cpu 2:50.43 total
```

After change 

```
# time ansible-inventory -i vmware_vm_inventory.yml --list
...
1.32s user 0.20s system 4% cpu 30.717 total
```

File: lib/ansible/plugins/inventory/vmware_vm_inventory.py [hunk 1]
```diff
@@ -379,6 +399,77 @@ def _populate_from_cache(self, source_data):
                     self._populate_host_vars([host], hostvars.get(host, {}), group)
                 self.inventory.add_child('all', group)
 
+    def _handle_error(self, message):
+        no_object_failure = self.get_option('no_object_failure') or 'silent'
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/vmware_vm_inventory.py:403]: With choices defined and the default set in the documentation I don't think `or 'silent'` is needed here.

File: lib/ansible/plugins/inventory/vmware_vm_inventory.py [hunk 2]
```diff
@@ -141,15 +208,15 @@ def __init__(self, hostname, username, password, port, validate_certs, with_tags
...
     def do_login(self):
         """
         Check requirements and do login
         """
         self.check_requirements()
-        self.content = self._login()
-        if self.with_tags:
-            self.rest_content = self._login_vapi()
+        self.si, self.content = self._login()
+        self.rest_content = self._login_vapi()
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/vmware_vm_inventory.py:219]: Is vSphere/vCloud required regardless of tags now, or was the original `if self.with_tags:` above this line just unnecessary?

File: lib/ansible/plugins/inventory/vmware_vm_inventory.py [hunk 3]
```diff
@@ -141,15 +208,15 @@ def __init__(self, hostname, username, password, port, validate_certs, with_tags
...
 
     def do_login(self):
         """
         Check requirements and do login
         """
         self.check_requirements()
-        self.content = self._login()
-        if self.with_tags:
-            self.rest_content = self._login_vapi()
+        self.si, self.content = self._login()
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/vmware_vm_inventory.py:218]: It doesn't look like the service instance is used anywhere after this.

**Review by s-hertel — APPROVED** [review]
This could use a changelog. It would also be nice if some of the new features had added tests, but it looks generally good to me. I don't have the resources to test so I just did a code review.

**Comment by goneri** [comment]
Hi @Akasurde,

I'm surprised because I actually see a performance drop:

- without the patch: 2678ms, 2788ms, 2465ms
- with the patch: 3404ms, 3247ms, 3261ms

Any idea what's going on?

**Comment by gparent** [comment]
I thought I had commented on this, I apologize. In my case the performance improved. I went from being unable to iterate my infrastructure to being able to assess about a hundred VM in a reasonable time once a folder was specified. I had taken numbers but it seems I never got to pasting them here.

**Comment by goneri** [comment]
Hi!

Thank you for your PR. Since March 2, 2020, the Ansible team does not accept any new changes on the VMware modules on the core `ansible/ansible` repository.
The development continue but in the new VMware collection repository: https://github.com/ansible-collections/vmware
Could you please reopen the PR there?

Links:
- https://groups.google.com/forum/#!msg/ansible-devel/6UAfogEDtG8/wMTgxBGWCgAJ
- https://github.com/ansible-collections/overview/blob/master/README.rst

**Comment by Akasurde** [comment]
Superseded by https://github.com/ansible-collections/vmware/pull/45/

**Pull request closed without merging**